### PR TITLE
E2E tests for kfto pytorch image

### PR DESCRIPTION
**Pull request by Fiona-Waters**
[RHOAIENG-14536](https://issues.redhat.com/browse/RHOAIENG-14536)
[RHOAIENG-14537](https://issues.redhat.com/browse/RHOAIENG-14537)
[RHOAIENG-14568](https://issues.redhat.com/browse/RHOAIENG-14568)

## Description
 - Added an E2E test to cover usage of the CUDA and ROCm training images.
 - Included a [python script taken from the training-operator](https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/trainer/hf_llm_training.py) for which I have included the apache license and a reference to the location of the original script. I have noted in a comment where the script has been altered.
 - Added environment variables so that the test can be used for both CUDA and ROCm images.

## How Has This Been Tested?

I have run the test with the CUDA image on a rosa cluster.  (Plan to test ROCm image next on AMD cluster.)

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: tests/kfto/core/environment.go [hunk 1]
```diff
@@ -42,6 +42,10 @@ const (
 	sleepImageEnvVar = "SLEEP_IMAGE"
 	// The environment variable specifying s3 bucket folder path used to store model
 	storageBucketModelPath = "AWS_STORAGE_BUCKET_MODEL_PATH"
+	// The environment variable for the CUDA Training image
```
Comment (oksanabaza) [inline tests/kfto/core/environment.go:45]: just a very small nit
```suggestion
	// The environment variable for the CUDA training image
```

File: tests/kfto/core/kfto_training_test.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+func TestPytorchjobWithCuda(t *testing.T) {
+	test := With(t)
+	cudaBaseImage := GetCudaTrainingImage(test)
+	runKFTOPytorchJob(t, cudaBaseImage, 1)
+}
+
+func TestPytorchjobWithROCm(t *testing.T) {
+	test := With(t)
+	rocmBaseImage := GetROCmTrainingImage(test)
+	runKFTOPytorchJob(t, rocmBaseImage, 0)
```
Comment (oksanabaza) [inline tests/kfto/core/kfto_training_test.go:43]: should we use 1 GPU in this case as well?

File: tests/kfto/core/environment.go [hunk 3]
```diff
@@ -52,6 +56,22 @@ func GetFmsHfTuningImage(t Test) string {
...
+func GetCudaTrainingImage(t Test) string {
+	image, ok := os.LookupEnv(cudaTrainingImageEnvVar)
+	if !ok {
+		t.T().Fatalf("Expected environment variable %s not found, please use this environment variable to specify the cuda training image to be tested.", cudaTrainingImageEnvVar)
+	}
+	return image
+}
+
+func GetROCmTrainingImage(t Test) string {
+	image, ok := os.LookupEnv((rocmTrainingImageEnvVar))
```
Comment (astefanutti) [inline tests/kfto/core/environment.go:68]:
```suggestion
	image, ok := os.LookupEnv(rocmTrainingImageEnvVar)
```

File: tests/kfto/core/kfto_training_test.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+	. "github.com/project-codeflare/codeflare-common/support"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kftov1 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v1"
+)
+
+func TestPytorchjobWithCuda(t *testing.T) {
```
Comment (astefanutti) [inline tests/kfto/core/kfto_training_test.go:34]:
```suggestion
func TestPyTorchJobWithCuda(t *testing.T) {
```

File: tests/kfto/core/kfto_training_test.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	test := With(t)
+	cudaBaseImage := GetCudaTrainingImage(test)
+	runKFTOPytorchJob(t, cudaBaseImage, 1)
+}
+
+func TestPytorchjobWithROCm(t *testing.T) {
```
Comment (astefanutti) [inline tests/kfto/core/kfto_training_test.go:40]:
```suggestion
func TestPyTorchJobWithROCm(t *testing.T) {
```

File: tests/kfto/core/kfto_training_test.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+	test := With(t)
+	rocmBaseImage := GetROCmTrainingImage(test)
+	runKFTOPytorchJob(t, rocmBaseImage, 0)
+}
+
+func runKFTOPytorchJob(t *testing.T, image string, numGpus int) {
```
Comment (astefanutti) [inline tests/kfto/core/kfto_training_test.go:46]:
```suggestion
func runKFTOPyTorchJob(t *testing.T, image string, numGpus int) {
```

File: tests/kfto/core/kfto_training_test.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+		Spec: kftov1.PyTorchJobSpec{
+			PyTorchReplicaSpecs: map[kftov1.ReplicaType]*kftov1.ReplicaSpec{
+				"Master": {
+					Replicas:      Ptr(int32(1)),
+					RestartPolicy: "OnFailure",
+					Template: corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Tolerations: []corev1.Toleration{
+								{
+									Key:      "nvidia.com/gpu",
```
Comment (astefanutti) [inline tests/kfto/core/kfto_training_test.go:101]: It may need to be adapted for ROCm.

**Review by astefanutti — COMMENTED** [review]
This looks great! Just left few comments. It'll be good to go after that.

File: tests/kfto/core/environment.go (hunk 1 above)
Comment (Fiona-Waters) [reply tests/kfto/core/environment.go]: Updated, thank you :)

File: tests/kfto/core/kfto_training_test.go (hunk 2 above)
Comment (Fiona-Waters) [reply tests/kfto/core/kfto_training_test.go]: Yes, I have updated this. Thanks

File: tests/kfto/core/kfto_training_test.go (hunk 7 above)
Comment (Fiona-Waters) [reply tests/kfto/core/kfto_training_test.go]: Thank you. I've updated so the test can be used for both. Just setting up to run on AMD gpu now.

**Review by astefanutti — APPROVED** [review]
LGTM

**Pull request merged**